# Agents and Tools

Agents involve an LLM making decisions about which Actions to take, taking that Action, seeing an Observation, and repeating that until done.


When used correctly agents can be extremely powerful. In order to load agents, you should understand the following concepts:

- Tool: A function that performs a specific duty. This can be things like: Google Search, Database lookup, Python REPL, other chains.
- LLM: The language model powering the agent.
- Agent: The agent to use.


For example I have to travel from Dubai to Canada, I type this in ChatGPT

Give me two flight options from Dubai to Canada on September 1, 2024 | ChatGPT will not be able to answer because has knowledge till
September 2021

ChatGPT plus has Expedia Plugin, if we enable this plugin it will go to Expedia Plugin and will try to pull information about Flights & it will show the information

When we think about LLM. Many people think that it is just a knowledge engine, it has knowledge and it will try to give answer based on that knowledge but the knowledge is only limited to September 2021. The think that most people missout is that Large Lanaguage Model has a reasoning engine, and using that reasoning engine it can figure out when someone types this type of Question


Give me  two flight options from Dubai to Canada on September 1, 2024


As a human we go to Expedia as we have a reasoning engine in our brain.

LLM has a reasoning engine as well, so it will figure out the Source, Destination, Date and it will call Expedia Plugin and it will return response back.

#**How much is US GDP in 2024? plus 5**


# Agent will conenct with external tools and it will use LLM reasoning capabilities

# To access Google Search Results in Real Time we use serpapi

# serpapi and llm-math tool

In [ ]:
!pip install google-search-results

  Preparing metadata (setup.py) ... done
  Created wheel for google-search-results: filename=google_search_results-2.4.2-py3-none-any.whl size=32004 sha256=da43d692eddc939def3221af0c7c328670a3ac0638396d7a4e14911995c9a830
  Stored in directory: /root/.cache/pip/wheels/d3/b2/c3/03302d12bb44a2cdff3c9371f31b72c0c4e84b8d2285eeac53
Successfully built google-search-results


In [ ]:
!pip install langchain

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 975.5/975.5 kB 8.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 337.4/337.4 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.5/127.5 kB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.1/141.1 kB 14.5 MB/s eta 0:00:00


In [ ]:
!pip install -U langchain-community

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 5.5 MB/s eta 0:00:00


In [ ]:
%pip install -qU langchain-openai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.9/45.9 kB 1.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 328.3/328.3 kB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 7.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 5.0 MB/s eta 0:00:00


In [ ]:
import os

In [ ]:
from google.colab import userdata
SERPAPI_API_KEY=userdata.get('SERPAPI_API_KEY')
GOOGLE_API_KEY=userdata.get('GOOGLE_API_KEY')

In [ ]:
os.environ["SERPAPI_API_KEY"]=SERPAPI_API_KEY

In [ ]:
os.environ["GOOGLE_API_KEY"]=GOOGLE_API_KEY

In [ ]:
%pip install --upgrade --quiet  langchain-google-genai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 163.9/163.9 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 718.3/718.3 kB 9.6 MB/s eta 0:00:00


In [ ]:
from langchain_google_genai import ChatGoogleGenerativeAI

## To check if the gemini is working or not.

llm = ChatGoogleGenerativeAI(model="gemini-1.5-pro")
result = llm.invoke("Write a ballad about LangChain")
print(result.content)

In [ ]:
llm = ChatGoogleGenerativeAI(model="gemini-1.5-pro")

In [ ]:
from langchain.agents import AgentType,initialize_agent, load_tools

In [ ]:
from langchain_openai import ChatOpenAI

## If you want to use OpenAI

llm = ChatOpenAI(
    model="gpt-3.5-turbo",
    temperature=0)

In [ ]:
tool = load_tools(["serpapi","llm-math"],llm=llm)

In [ ]:
agent1 = initialize_agent(tool, llm, agent= AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose= True)

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning: The function `initialize_agent` was deprecated in LangChain 0.1.0 and will be removed in 0.3.0. Use Use new agent constructor methods like create_react_agent, create_json_agent, create_structured_chat_agent, etc. instead.
  warn_deprecated(


In [ ]:
agent1.run("How much is US GDP in 2024 plus 5?")

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:139: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(




> Entering new AgentExecutor chain...
Thought: I need to search what the US GDP is projected to be in 2024 and then add 5 to that number.
Action: Search
Action Input: US GDP projection 2024
Observation: From 2024 through 2028, GDP will increase at an average annual rate of 2.2% per year, 0.5 percentage points higher than the baseline forecast. This scenario also results in higher long-term potential for the economy at 2.3%, compared to 1.7% in the baseline.
Thought:Thought: I need to find the actual GDP number for 2024 from that information, I can't calculate it with what I have.
Action: Search
Action Input: US GDP 2024
Observation: Real gross domestic product (GDP) increased at an annual rate of 1.3 percent in the first quarter of 2024 (table 1), according to the "second" estimate released by the Bureau of Economic Analysis. In the fourth quarter of 2023, real GDP increased 3.4 percent.
Thought:Thought: I need to find the projected GDP for 2024, not the current rate. 
Action: Search

Thought: The top result gives a forecast in billions of US dollars. I can use that.
Action: Calculator
Action Input: 28176.6 + 5 


Observation: Answer: 28181.6
Thought:Thought: I now know the final answer
Final Answer: The projected US GDP in 2024 plus 5 is $28,181.6 billion. 


> Finished chain.


'The projected US GDP in 2024 plus 5 is $28,181.6 billion.'

In [ ]:
agent1.run("in which year indian team won its second world cup give me all the player name of the winning team??")



> Entering new AgentExecutor chain...
Thought: I need to search for information about the Indian cricket team's second world cup victory.
Action: Search
Action Input: indian cricket team second world cup year and team players 
Observation: ["The Indian team was led by Srinivas Venkataraghavan for the second consecutive world cup. India's first match was against reigning world champions West ...", 'Checkout team and player detaills of India Squad - ICC Cricket World Cup, 2011 Squad details on ESPNcricinfo.com.', 'India was the second team (after the West Indies) to win the World Cup, and the first to win the competition on home soil after winning it in 2011. India are ...', 'Indian players in the team of the T20 World Cup 2024: ; Rohit Sharma (Runs: 257, Average: 36.71, Strike-rate: 156.7, Fifties: 3) ; Suryakumar ...', 'India will be led by Rohit Sharma as they aim to bag their second T20 World Cup trophy. Notable players in the squad include Virat Kohli and', "Here's a summary of th

'India won its second world cup in 2011. The players were Mahendra Singh Dhoni (captain and wicketkeeper), Virender Sehwag (vice-captain), Ravichandran Ashwin, Piyush Chawla, Gautam Gambhir, Zaheer Khan, Virat Kohli, Ashish Nehra, Munaf Patel, Yusuf Pathan, Suresh Raina, Harbhajan Singh, Yuvraj Singh, Sreesanth, and Sachin Tendulkar.'

In [ ]:
agent1.run("How much is india's GDP in 2024 plus 5?")



> Entering new AgentExecutor chain...
Thought: I need to search for India's GDP in 2024 and then add 5 to it.
Action: Search
Action Input: India's GDP in 2024
Observation: {'type': 'organic_result', 'title': 'Economy of India', 'source': 'Wikipedia'}
Thought:Thought: I need to search for India's GDP in 2024 and then add 5 to it.
Action: Search
Action Input: India's GDP in 2024
Observation: {'type': 'organic_result', 'title': 'Economy of India', 'source': 'Wikipedia'}
Thought:Thought: I need to find India's GDP in 2024 and then add 5 to it.
Action: Search
Action Input: India's GDP in 2024
Observation: {'type': 'organic_result', 'title': 'Economy of India', 'source': 'Wikipedia'}
Thought:Thought: I need to find India's GDP in 2024 and then add 5 to it.
Action: Search
Action Input: India's GDP 2024
Observation: $14.59 trillion (PPP; 2024 est.)
Thought:Thought: I need to extract the numerical value from the observation and add 5 to it.
Action: Calculator
Action Input: 14.59 trillion + 5


"India's GDP in 2024 plus 5 is 14,590,000,000,005."

In [ ]:
agent1.run("who won the india's 2024 general election?")



> Entering new AgentExecutor chain...
Thought: I need to search for the results of India's 2024 general election.
Action: Search
Action Input: India 2024 general election results
Observation: [{'title': 'UK election results 2024: Why do some popular parties win so few seats?', 'link': 'https://www.aljazeera.com/news/2024/7/5/uk-general-election-2024-why-do-some-popular-parties-win-so-few-seats', 'source': 'Al Jazeera', 'date': '1 day ago', 'thumbnail': 'https://serpapi.com/searches/668974e23bea7bb292e38bbd/images/088ecbb36d17e2e9b2af24e6cbd99563d9dd4379a664f0a3.jpeg'}, {'title': 'UK Election: Record Number Of Indian-Origin MPs Elected To British Parliament', 'link': 'https://www.ndtv.com/world-news/2024-uk-election-result-record-number-of-indian-origin-mps-elected-to-british-parliament-6039751', 'source': 'NDTV', 'date': '1 day ago', 'thumbnail': 'https://serpapi.com/searches/668974e23bea7bb292e38bbd/images/088ecbb36d17e2e9c57d25cdee3c886570296e589812c4db.jpeg'}, {'title': 'India ele


Observation: [{'title': 'UK election results 2024: Why do some popular parties win so few seats?', 'link': 'https://www.aljazeera.com/news/2024/7/5/uk-general-election-2024-why-do-some-popular-parties-win-so-few-seats', 'source': 'Al Jazeera', 'date': '1 day ago', 'thumbnail': 'https://serpapi.com/searches/668974e866440ad81cc57937/images/4477abfcefdb5d286d0211a312bf5fefc549d22ca138476d.jpeg'}, {'title': 'UK General Election 2024: All you need to know about the polling today', 'link': 'https://indianexpress.com/article/world/uk-general-election-2024-all-you-need-to-know-about-polls-day-9431030/', 'source': 'The Indian Express', 'date': '2 days ago', 'thumbnail': 'https://serpapi.com/searches/668974e866440ad81cc57937/images/4477abfcefdb5d28f64a17d290a633ec3feafe06441703c1.jpeg'}, {'title': 'India election 2024: Why Modi failed to win outright majority', 'link': 'https://www.bbc.com/news/articles/c977g8gl5q2o', 'source': 'BBC', 'date': '1 month ago', 'thumbnail': 'https://serpapi.com/sea


Observation: ['General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', "Prime Minister Narendra Modi's National Democratic Alliance coalition has won a majority in parliament, according to official results from ...", 'The governing Bharatiya Janata Party (BJP) of Prime Minister Narendra Modi won 240 seats, falling short of the 272-mark that signifies a ...', "Prime Minister Narendra Modi will serve a third term in office after India's election authority confirmed the National Democratic Alliance coalition headed ...", "Modi's Bharatiya Janata Party, or B.J.P., would still win by far the most parliamentary seats, and that he would become only the second Indian ...", 'The Bharatiya Janata Party, together with its National Democratic Alliance, has won a majority.', 'Prime Minister Narendra Modi and his Bharatiya Janata Party (BJP) won the most seats for the third consecutive time, but h

Thought: I see that the National Democratic Alliance (NDA) won a majority in India's 2024 general election.
Final Answer: The National Democratic Alliance (NDA) won India's 2024 general election. 


> Finished chain.


"The National Democratic Alliance (NDA) won India's 2024 general election."

In [ ]:
agent1.run("how much seat BJP won in india's general election in 2024 plus 10000?")



> Entering new AgentExecutor chain...
Thought: I need to find out how many seats BJP won in the 2024 Indian general election and then add 10000 to that number.
Action: Search
Action Input: How many seats did BJP win in India's 2024 general election?
Observation: ["Prime Minister Narendra Modi's National Democratic Alliance coalition has won a majority in parliament, according to official results from ...", 'The governing Bharatiya Janata Party (BJP) of Prime Minister Narendra Modi won 240 seats, falling short of the 272-mark that signifies a ...', 'General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', 'Indian PM Narendra Modi has won a third consecutive term in a much tighter general election than anticipated. His Bharatiya Janata Party (BJP) ...', 'The Bharatiya Janata Party, together with its National Democratic Alliance, has won a majority.', 'Election results show Indian Prime

'The BJP won 240 seats in the 2024 Indian general election. Adding 10000 to that number gives us 10240.'

In [ ]:
!pip install wikipedia

  Preparing metadata (setup.py) ... done
  Created wheel for wikipedia: filename=wikipedia-1.4.0-py3-none-any.whl size=11680 sha256=7d5124ea393a44b9c8c0341e68d32afff5bd6020e59115aea06b151011fe98ed
  Stored in directory: /root/.cache/pip/wheels/5e/b6/c5/93f3dec388ae76edc830cb42901bb0232504dfc0df02fc50de
Successfully built wikipedia


In [ ]:
tool = load_tools(["wikipedia","llm-math"],llm=llm)

In [ ]:
agent = initialize_agent(
    tool,
    llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    verbose=True,

)

In [ ]:
agent.run("in which year first indian film released?")



> Entering new AgentExecutor chain...
Thought: I need to search for the first Indian film on Wikipedia.
Action: wikipedia
Action Input: first Indian film


Observation: Page: Cinema of India
Summary: The Cinema of India, consisting of motion pictures made by the Indian film industry, has had a large effect on world cinema since the second half of the 20th century. Indian cinema is made up of various film industries, including Hindi cinema, which makes motion pictures in the Hindi language and is one of the biggest film industries in the country. In 2021, Telugu cinema emerged as the largest film industry in India in terms of box office. Major centers of film production across the country include Mumbai, Hyderabad, Chennai, Kolkata, Kochi, Bangalore, Bhubaneswar-Cuttack, and Guwahati. For a number of years, the Indian film industry has ranked first in the world in terms of annual film output. In 2022, Indian cinema earned ₹15,000 crore ($1.7955 billion) at the box-office. Ramoji Film City located in Hyderabad, India is certified by the Guinness World Records as the largest film studio complex in the world measuring over 1,666 acres (674 h

'1913'

In [ ]:
agent.run("give me a detail analysis of indian election 2024")



> Entering new AgentExecutor chain...
Thought: I need more information to provide a detailed analysis of the Indian election in 2024. It's too early for concrete data. I can give some general information about Indian elections. 
Action: wikipedia
Action Input: Indian general elections
Observation: Page: 2024 Indian general election
Summary: General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the result was declared on 4 June to form the 18th Lok Sabha. On 7 June 2024, Prime Minister Narendra Modi confirmed the support of 293 MPs to Droupadi Murmu, the president of India. This marked Modi's third term as prime minister and his first time heading a coalition government, with the Telugu Desam Party of Andhra Pradesh and Janata Dal (United) of Bihar emerging as two main allies.
More than 968 million people out of a population of 1.4 billion people were eligible to vote, equivalent to 70 perce

"The 2024 Indian general election took place from April 19th to June 1st, 2024, with Narendra Modi's BJP-led NDA securing victory and Modi elected for a third term as Prime Minister. While the NDA won, forming a coalition government, the BJP itself lost its single-party majority, securing 240 seats compared to 303 in 2019. The opposition INDIA coalition won 234 seats, making significant gains."

The 2024 Indian general election was held in seven phases from 19 April to 1 June 2024 to elect all 543 members of the Lok Sabha. Narendra Modi secured his third term as prime minister, leading a coalition government with the support of 293 MPs. The election saw a record number of eligible voters and high participation, with the BJP losing its majority in the Lok Sabha.